tokio/src/runtime/local_runtime/options.rs

```diff
@@ -1,18 +1,142 @@
+    /// # pub fn main() {
+    /// let (tx, rx) = std::sync::mpsc::channel();
+    /// let mut opts = LocalOptions::default();
+    /// opts.on_thread_park(move || match rx.recv() {
+    ///    Ok(x) => println!("Received from channel: {}", x),
+    ///    Err(e) => println!("Error receiving from channel: {}", e),
+    /// });
+    ///
+    /// let runtime = Builder::new_current_thread()
+    ///     .enable_time()
+    ///     .build_local(opts)
+    ///     .unwrap();
+    ///
+    /// runtime.block_on(async {
+    ///     tokio::task::spawn_local(async move {
+    ///         tx.send(42).unwrap();
+    ///     });
+    ///     tokio::time::sleep(std::time::Duration::from_millis(1)).await;
+    /// })
+    /// }
+    /// ```
+    pub fn on_thread_park<F>(&mut self, f: F) -> &mut Self
+    where
+        F: Fn() + 'static,
+    {
+        self.before_park = Some(std::sync::Arc::new(to_send_sync(f)));
+        self
+    }
+
+    /// Executes function `f` just after a thread unparks (starts executing tasks).
+    ///
+    /// This is intended for bookkeeping and monitoring use cases; note that work
+    /// in this callback will increase latencies when the application has allowed one or
+    /// more runtime threads to go idle.
+    ///
+    /// This differs from the `Builder::on_thread_unpark` method in that it accepts a non Send + Sync
+    /// closure.
+    ///
+    /// Note: There can only be one unpark callback for a runtime; calling this function
+    /// more than once replaces the last callback defined, rather than adding to it.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use tokio::runtime::{Builder, LocalOptions};
+    /// # pub fn main() {
+    /// let (tx, rx) = std::sync::mpsc::channel();
+    /// let mut opts = LocalOptions::default();
+    /// opts.on_thread_unpark(move || match rx.recv() {
+    ///    Ok(x) => println!("Received from channel: {}", x),
+    ///    Err(e) => println!("Error receiving from channel: {}", e),
+    /// });
+    ///
+    /// let runtime = Builder::new_current_thread()
+    ///     .enable_time()
+    ///     .build_local(opts)
+    ///     .unwrap();
+    ///
+    /// runtime.block_on(async {
+    ///     tokio::task::spawn_local(async move {
+    ///         tx.send(42).unwrap();
+    ///     });
+    ///     tokio::time::sleep(std::time::Duration::from_millis(1)).await;
+    /// })
+    /// }
+    /// ```
+    pub fn on_thread_unpark<F>(&mut self, f: F) -> &mut Self
+    where
+        F: Fn() + 'static,
+    {
+        self.after_unpark = Some(std::sync::Arc::new(to_send_sync(f)));
+        self
+    }
+}
+
+// A wrapper type to allow non-Send + Sync closures to be used in a Send + Sync context.
+// This is specifically used for executing callbacks when using a `LocalRuntime`, since the
+// callbacks will never be sent across threads and thus do not need to be Send or Sync.
+struct UnsafeSendSync<T>(T);
+unsafe impl<T> Send for UnsafeSendSync<T> {}
```

Comment:
shouldn't this be `unsafe impl<T: Send> Send for UnsafeSendSync<T> {}`
and `unsafe impl<T: Send + Sync> Sync for UnsafeSendSync<T> {}`